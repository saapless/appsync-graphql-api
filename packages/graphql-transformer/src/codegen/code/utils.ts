/* eslint-disable security/detect-object-injection */

import { PropertyValue, tc } from "./ast";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}
