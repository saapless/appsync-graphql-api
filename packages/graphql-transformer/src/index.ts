export * from "./transformer";
export { definitionFromSource } from "./utils";
