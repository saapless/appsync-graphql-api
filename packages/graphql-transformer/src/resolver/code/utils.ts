export function wrap(start: string, value: string, end = "") {
  return `${start}${value}${end}`;
}

export function indent(value: string) {
  return value
    .split("\n")
    .map((line) => `  ${line}`)
    .join("\n");
}

export function join<T extends (string | string[])[]>(separator = "", ...statements: T) {
  return statements.flat().filter(Boolean).join(separator);
}

export function block<T extends string[]>(...value: T) {
  return wrap("{\n", indent(join("\n", value)).trimEnd(), "\n}\n");
}

export function expression(value: string) {
  return wrap("(", value, ")");
}

export function statement(value: string) {
  return wrap("", value, `${value.endsWith(";") ? "" : ";"}\n`);
}

export function returnStatement(value: string) {
  return wrap("return ", statement(value));
}
