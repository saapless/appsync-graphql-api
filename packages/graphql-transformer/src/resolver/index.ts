export { ResolverBase } from "./ResolverBase";
export { FieldResolver } from "./FieldResolver";
export { FunctionResolver } from "./FunctionResolver";
