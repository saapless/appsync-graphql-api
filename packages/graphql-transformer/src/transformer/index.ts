export { GraphQLTransformer, type GraphQLTransformerOptions } from "./GraphQLTransformer";
export { createTransformer } from "./createTransformer";
