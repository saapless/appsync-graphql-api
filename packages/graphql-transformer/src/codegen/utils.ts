/* eslint-disable security/detect-object-injection */

import { KeyValue } from "../utils/types";
import { PropertyValue, tc } from "./code";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}

export function keyValue<T extends string | number>(obj: KeyValue<T>) {
  if (obj.ref) return tc.chain("ctx", obj.ref);
  if (obj.eq) return typeof obj.eq === "string" ? tc.str(obj.eq) : tc.num(obj.eq);
  throw new Error("Invalid key value");
}
