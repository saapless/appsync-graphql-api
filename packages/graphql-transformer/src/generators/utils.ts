/* eslint-disable security/detect-object-injection */

import { FieldLoaderDescriptor, Key, KeyValue, LoaderDescriptor } from "../utils/types";
import { PropertyValue, tc } from "../codegen";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}

export function keyValue<T extends string | number>(obj: KeyValue<T>) {
  if (obj.ref) {
    // `ctx.source` is typed as optional in Context

    if (obj.ref.startsWith("source")) {
      obj.ref = obj.ref.replace("source", "source?");
    }

    return tc.chain("ctx", obj.ref);
  }
  if (obj.eq) return typeof obj.eq === "string" ? tc.str(obj.eq) : tc.num(obj.eq);
  throw new Error("Invalid key value");
}

export function isFieldLoader(descriptor: LoaderDescriptor): descriptor is FieldLoaderDescriptor {
  return Object.hasOwn(descriptor, "fieldName") && Object.hasOwn(descriptor, "typeName");
}

export function parseKey(key: Key) {
  return tc.obj(
    Object.entries(key).reduce(
      (agg, [name, op]) => {
        if (op.ref || op.eq) {
          Object.assign(agg, { [name]: keyValue(op) });
        } else if (op.le) {
          Object.assign(agg, { [name]: tc.obj({ le: keyValue(op.le) }) });
        } else if (op.lt) {
          Object.assign(agg, { [name]: tc.obj({ lt: keyValue(op.lt) }) });
        } else if (op.ge) {
          Object.assign(agg, { [name]: tc.obj({ ge: keyValue(op.ge) }) });
        } else if (op.gt) {
          Object.assign(agg, { [name]: tc.obj({ gt: keyValue(op.gt) }) });
        } else if (op.between) {
          Object.assign(agg, {
            [name]: tc.obj({ between: tc.arr(...op.between.map((i) => keyValue(i))) }),
          });
        } else if (op.beginsWith) {
          Object.assign(agg, { [name]: tc.obj({ beginsWith: keyValue(op.beginsWith) }) });
        }
        return agg;
      },
      {} as Record<string, PropertyValue>
    )
  );
}
