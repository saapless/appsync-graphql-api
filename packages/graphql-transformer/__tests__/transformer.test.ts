import { readFileSync } from "node:fs";
import { join } from "node:path";
import { createTransformer, GraphQLTransformer } from "../src/transformer";
import { FieldResolver } from "../src/resolver";
import {
  DirectiveDefinitionNode,
  EnumNode,
  FieldNode,
  NamedTypeNode,
  ObjectNode,
  ScalarNode,
} from "../src/parser";
import { SchemaValidationError } from "../src/utils/errors";

// eslint-disable-next-line security/detect-non-literal-fs-filename
const schema = readFileSync(join(__dirname, "./schema.graphql"), "utf-8");

describe("GraphQLTransformer", () => {
  const transformer = createTransformer({ definition: schema });

  describe("createTransformer factory", () => {
    it("throws if empty definition", () => {
      expect(() => createTransformer({ definition: "" })).toThrow();
    });

    it("creates new GraphQLTransformer instance", () => {
      expect(transformer).toBeDefined();
      expect(transformer).toBeInstanceOf(GraphQLTransformer);
    });

    it("adds default plugins list", () => {
      expect(transformer.plugins).toHaveLength(5);
    });
  });

  describe("given invalid schema", () => {
    it("throws SchemaValidationError", () => {
      const transformer = createTransformer({
        definition: schema.replace("type Viewer", "type Viewer2"),
      });

      expect(() => transformer.transform()).toThrow(SchemaValidationError);
    });
  });

  describe("runs schema transformations", () => {
    const result = transformer.transform();

    it("creates valid schema", () => {
      const schema = result.document.print();
      expect(schema).toMatchSnapshot();
    });

    it("copies extended fields", () => {
      const viewerType = result.document.getNode("Viewer") as ObjectNode;

      expect(viewerType.getField("user")).toBeInstanceOf(FieldNode);
      expect(viewerType.getField("tasks")).toBeInstanceOf(FieldNode);
    });

    describe("AWSTypesPlugin transformations", () => {
      it("adds AWS specific scalars", () => {
        const scalars = [
          "AWSDate",
          "AWSTime",
          "AWSDateTime",
          "AWSTimestamp",
          "AWSEmail",
          "AWSJSON",
          "AWSURL",
          "AWSPhone",
          "AWSIPAddress",
        ];

        scalars.forEach((scalar) => {
          expect(result.document.getNode(scalar)).toBeInstanceOf(ScalarNode);
        });
      });

      it("adds AWS specific directives", () => {
        const directives = [
          "aws_api_key",
          "aws_auth",
          "aws_cognito_user_pools",
          "aws_lambda",
          "aws_oidc",
          "aws_subscribe",
        ];

        directives.forEach((directive) => {
          expect(result.document.getNode(directive)).toBeInstanceOf(DirectiveDefinitionNode);
        });
      });
    });

    describe("NodeInterfacePlugin trnsformations", () => {
      it("adds valid `node` field to Query", () => {
        const nodeField = (result.document.getNode("Query") as ObjectNode)?.getField("node");
        const nodeFieldTypename = (nodeField?.type as NamedTypeNode).name;

        expect(nodeField).toBeDefined();
        expect(nodeField).toBeInstanceOf(FieldNode);
        expect(nodeField?.hasArgument("id")).toStrictEqual(true);

        expect(nodeField?.type).toBeInstanceOf(NamedTypeNode);
        expect(nodeFieldTypename).toStrictEqual("Node");
      });

      it("adds Query.node resolver", () => {
        const nodeResolver = result.resolvers.get("Query.node");

        expect(nodeResolver).toBeInstanceOf(FieldResolver);
        expect(nodeResolver?.print()).toMatchSnapshot();
      });

      it("extends models with Node interface", () => {
        const userNode = result.document.getNode("User") as ObjectNode;
        const taskNode = result.document.getNode("Task") as ObjectNode;

        expect(userNode.hasInterface("Node")).toBeTruthy();
        expect(taskNode.hasInterface("Node")).toBeTruthy();
      });
    });

    describe("ModelPlugin", () => {
      it("adds @model directive definition", () => {
        const modelDirective = result.document.getNode("model");
        const operationsEnum = result.document.getNode("ModelOperation");

        expect(modelDirective).toBeInstanceOf(DirectiveDefinitionNode);
        expect(operationsEnum).toBeInstanceOf(EnumNode);
      });

      it("created query fields for model", () => {
        const queryNode = result.document.getQueryNode();

        expect(queryNode.getField("getUser")).toBeDefined();
        expect(queryNode.getField("listUsers")).toBeDefined();
        expect(queryNode.getField("getUser")).toBeInstanceOf(FieldNode);
        expect(queryNode.getField("listUsers")).toBeInstanceOf(FieldNode);

        expect(queryNode.getField("getTask")).not.toBeDefined();
        expect(queryNode.getField("listTasks")).not.toBeDefined();
      });
      it("created mutation fields for model", () => {
        const mutationNode = result.document.getMutationNode();

        expect(mutationNode.getField("createUser")).toBeDefined();
        expect(mutationNode.getField("updateUser")).toBeDefined();
        expect(mutationNode.getField("deleteUser")).toBeDefined();
        expect(mutationNode.getField("createUser")).toBeInstanceOf(FieldNode);
        expect(mutationNode.getField("updateUser")).toBeInstanceOf(FieldNode);
        expect(mutationNode.getField("deleteUser")).toBeInstanceOf(FieldNode);

        expect(mutationNode.getField("upsertTask")).toBeDefined();
        expect(mutationNode.getField("deleteTask")).toBeDefined();
        expect(mutationNode.getField("upsertTask")).toBeInstanceOf(FieldNode);
        expect(mutationNode.getField("deleteTask")).toBeInstanceOf(FieldNode);
      });

      it("created operation inputs & types", () => {
        expect(result.document.getNode("UserFilterInput")).toBeDefined();
        expect(result.document.getNode("CreateUserInput")).toBeDefined();
        expect(result.document.getNode("UpdateUserInput")).toBeDefined();
        expect(result.document.getNode("DeleteUserInput")).toBeDefined();
        expect(result.document.getNode("UpsertTaskInput")).toBeDefined();
        expect(result.document.getNode("DeleteTaskInput")).toBeDefined();
      });

      it("created operation resolvers", () => {
        expect(result.resolvers.get("Query.getUser")).toBeDefined();
        expect(result.resolvers.get("Query.listUsers")).toBeDefined();

        expect(result.resolvers.get("Mutation.createUser")).toBeDefined();
        expect(result.resolvers.get("Mutation.updateUser")).toBeDefined();
        expect(result.resolvers.get("Mutation.deleteUser")).toBeDefined();
        expect(result.resolvers.get("Mutation.upsertTask")).toBeDefined();
        expect(result.resolvers.get("Mutation.deleteTask")).toBeDefined();
      });
    });
  });
});
