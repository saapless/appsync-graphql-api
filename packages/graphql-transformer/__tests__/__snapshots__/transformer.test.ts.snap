// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLTransformer runs schema transformations respond with valid output 1`] = `
{
  "fieldResolvers": {
    "Mutation.createUser": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "createUser",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.deleteFile": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "deleteFile",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.deleteLabel": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "deleteLabel",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.deleteTask": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "deleteTask",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.deleteTimeTracker": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "deleteTimeTracker",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.deleteUser": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "deleteUser",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.updateUser": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "updateUser",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.upsertFile": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "upsertFile",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.upsertLabel": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "upsertLabel",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.upsertTask": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "upsertTask",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Mutation.upsertTimeTracker": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "upsertTimeTracker",
      "pipelineFunctions": undefined,
      "typeName": "Mutation",
    },
    "Query.getUser": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "getUser",
      "pipelineFunctions": undefined,
      "typeName": "Query",
    },
    "Query.listUsers": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "listUsers",
      "pipelineFunctions": undefined,
      "typeName": "Query",
    },
    "Query.node": {
      "code": "import { util } from "@aws-appsync/utils";
import { get } from "@aws-appsync/utils/dynamodb";
export function request(ctx) {
  return get({ key: { id: ctx.args.id } })
}

export function response(ctx) {
  return ctx.result
}
",
      "dataSource": undefined,
      "fieldName": "node",
      "pipelineFunctions": undefined,
      "typeName": "Query",
    },
    "Task.artifacts": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "artifacts",
      "pipelineFunctions": undefined,
      "typeName": "Task",
    },
    "Task.labels": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "labels",
      "pipelineFunctions": undefined,
      "typeName": "Task",
    },
    "Task.occurrences": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "occurrences",
      "pipelineFunctions": undefined,
      "typeName": "Task",
    },
    "Task.subtasks": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "subtasks",
      "pipelineFunctions": undefined,
      "typeName": "Task",
    },
    "Viewer.labels": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "labels",
      "pipelineFunctions": undefined,
      "typeName": "Viewer",
    },
    "Viewer.tasks": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "tasks",
      "pipelineFunctions": undefined,
      "typeName": "Viewer",
    },
    "Viewer.user": {
      "code": "export function request(ctx) {

}

export function response(ctx) {
  const { error, result } = ctx
  if (error) {
    util.error(error.message, erorr.type)
  }
  
  return result
}
",
      "dataSource": undefined,
      "fieldName": "user",
      "pipelineFunctions": undefined,
      "typeName": "Viewer",
    },
  },
  "pipelineFunctions": {},
  "schema": "type Viewer {
  user: User @connection(relation: oneOne, key: "userId", ref: "source.userId", index: null)
  tasks(filter: TaskFilterInput, first: Int, after: String, sort: [ModelSortDirection!]): TaskConnection! @edges
  labels(filter: LabelFilterInput, first: Int, after: String, sort: [ModelSortDirection!]): LabelConnection! @edges
  userId: ID @writeonly
}

interface Node {
  id: ID!
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: Int
  _deleted: Boolean @readonly
}

enum UserStatus {
  ACTIVE
  DISABLED
  SUSPENDED
}

type User implements Node @model {
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail
  picture: AWSURL
  status: UserStatus @readonly
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: Int
  _deleted: Boolean @readonly
}

type Task implements Node @model(operations: [upsert, delete]) {
  id: ID!
  title: String
  content: AWSJSON
  schedule: Schedule
  rrule: RRule
  recurrenceId: ID
  occurrenceId: String
  labels(filter: LabelFilterInput, first: Int, after: String, sort: [ModelSortDirection!]): LabelConnection! @connection(relation: "manyMany")
  subtasks(filter: TaskFilterInput, first: Int, after: String, sort: [ModelSortDirection!]): TaskConnection! @edges
  artifacts(first: Int, after: String, sort: [ModelSortDirection!]): ArtifactConnection! @edges
  occurrences(filter: TaskFilterInput, first: Int, after: String, sort: [ModelSortDirection!]): TaskConnection! @edges
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: Int
  _deleted: Boolean @readonly
}

type RRule {
  dtStart: AWSDateTime
  until: AWSDateTime
  ruleStr: String
}

type Schedule {
  startDate: DateTimeZone
  duration: String
  dueDate: DateTimeZone
}

type DateTimeZone {
  date: AWSDateTime
  timezone: String
}

type Label implements Node @model(operations: [upsert, delete]) {
  id: ID!
  name: String
  color: String
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: Int
  _deleted: Boolean @readonly
}

type File implements Node @model(operations: [upsert, delete]) {
  name: String
  size: Int
  url: AWSURL
  mimeType: String
  id: ID!
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: Int
  _deleted: Boolean @readonly
}

type TimeTracker implements Node @model(operations: [upsert, delete]) {
  duration: String
  logs: [TimeLog!]
  id: ID!
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: Int
  _deleted: Boolean @readonly
}

type TimeLog {
  action: TimeLogAction!
  timestamp: AWSTimestamp!
}

enum TimeLogAction {
  START
  PAUSE
  END
  LOG
}

union Artifact = File | TimeTracker

type Query {
  viewer: Viewer!
  node(id: ID!): Node
  getUser(id: ID!): User
  listUsers: User @connection(relation: oneMany, key: "__typename", index: "byTypename")
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int!]
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String!]
  contains: String
  notContains: String
  between: [String!]
  beginsWith: String
  attributeExists: Boolean
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  in: [Int!]
  between: [Int!]
  attributeExists: Boolean
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  in: [Float!]
  between: [Float!]
  attributeExists: Boolean
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
}

input ModelIDInput {
  ne: ID
  eq: ID
  in: [ID!]
  attributeExists: Boolean
}

input ModelListInput {
  contains: String
  notContains: String
  size: ModelSizeInput
}

enum SortDirection {
  ASC
  DESC
}

enum ConnectionRelationType {
  oneOne
  oneMany
  manyMany
}

directive @connection(relation: ConnectionRelationType, key: String, ref: String, index: String) on FIELD_DEFINITION | OBJECT

directive @node(key: String, ref: String, index: String) on FIELD_DEFINITION

directive @edges(key: String, ref: String, index: String) on FIELD_DEFINITION

directive @ignoreConnection on FIELD_DEFINITION

type PageInfo {
  hasNextPage: Boolean
  hasPreviousPage: Boolean
  startCursor: String
  endCursor: String
}

directive @auth on OBJECT | FIELD_DEFINITION | INTERFACE

directive @dataSource(source: String!) on OBJECT

type TaskConnection {
  edges: [TaskEdge!]!
  pageInfo: PageInfo!
}

type TaskEdge {
  cursor: String
  node: Task
}

input TaskFilterInput {
  id: ModelIDInput
  title: ModelStringInput
  content: ModelStringInput
  recurrenceId: ModelIDInput
  occurrenceId: ModelStringInput
  and: [TaskFilterInput]
  or: [TaskFilterInput]
  not: TaskFilterInput
}

type LabelConnection {
  edges: [LabelEdge!]!
  pageInfo: PageInfo!
}

type LabelEdge {
  cursor: String
  node: Label
}

input LabelFilterInput {
  id: ModelIDInput
  name: ModelStringInput
  color: ModelStringInput
  and: [LabelFilterInput]
  or: [LabelFilterInput]
  not: LabelFilterInput
}

input CreateUserInput {
  id: id
  firstName: String
  lastName: String
  email: AWSEmail
  picture: AWSURL
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: _version
}

type Mutation {
  createUser(input: CreateUserInput!): User
  updateUser(input: UpdateUserInput!): User
  deleteUser(input: DeleteUserInput!): User
  upsertTask(input: UpsertTaskInput!): Task
  deleteTask(input: DeleteTaskInput!): Task
  upsertLabel(input: UpsertLabelInput!): Label
  deleteLabel(input: DeleteLabelInput!): Label
  upsertFile(input: UpsertFileInput!): File
  deleteFile(input: DeleteFileInput!): File
  upsertTimeTracker(input: UpsertTimeTrackerInput!): TimeTracker
  deleteTimeTracker(input: DeleteTimeTrackerInput!): TimeTracker
}

input UpdateUserInput {
  id: id
  firstName: String
  lastName: String
  email: AWSEmail
  picture: AWSURL
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: _version
}

input DeleteUserInput {
  id: ID!
  _version: Int!
}

input DateTimeZoneInput {
  date: AWSDateTime
  timezone: String
}

input ScheduleInput {
  startDate: DateTimeZoneInput
  duration: String
  dueDate: DateTimeZoneInput
}

input RRuleInput {
  dtStart: AWSDateTime
  until: AWSDateTime
  ruleStr: String
}

input UpsertTaskInput {
  id: id
  title: String
  content: AWSJSON
  schedule: ScheduleInput
  rrule: RRuleInput
  recurrenceId: ID
  occurrenceId: String
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: _version
}

input DeleteTaskInput {
  id: ID!
  _version: Int!
}

type ArtifactConnection {
  edges: [ArtifactEdge!]!
  pageInfo: PageInfo!
}

type ArtifactEdge {
  cursor: String
  node: Artifact
}

input UpsertLabelInput {
  id: id
  name: String
  color: String
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: _version
}

input DeleteLabelInput {
  id: ID!
  _version: Int!
}

input UpsertFileInput {
  name: String
  size: Int
  url: AWSURL
  mimeType: String
  id: id
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: _version
}

input DeleteFileInput {
  id: ID!
  _version: Int!
}

input TimeLogInput {
  action: TimeLogAction
  timestamp: AWSTimestamp
}

input UpsertTimeTrackerInput {
  duration: String
  logs: TimeLogInput
  id: id
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  _version: _version
}

input DeleteTimeTrackerInput {
  id: ID!
  _version: Int!
}",
}
`;
